### ISSUE TITLE
basic transfer adapter tries to resume already complete file

### ISSUE THREAD

**Issue by robinst** [git-lfs/git-lfs#3172]
Hi. We've seen this in a recent support case. I'm not sure how it got into that state, but it looks like the "basic" transfer adapter will try to resume an already-completed file and fail. Log:

```
08:52:24.897474 trace git-lfs: HTTP: {"objects": [{"oid": "63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073", "actions": {"download": {"header": {"X-Client-ID": "...", "Authorization": "Bearer ..."}, "href": "https://api.media.atlassian.com/file/4d1155f8-7d31-4549-9dcd-851499a3b1d9/binary"}}, "size": 4707098120}]}
08:52:24.897474 trace git-lfs: tq: starting transfer adapter "basic"
08:52:37.884554 trace git-lfs: xfer: Attempting to resume download of "63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073" from byte 4707098120
08:52:37.893529 trace git-lfs: HTTP: GET https://api.media.atlassian.com/file/4d1155f8-7d31-4549-9dcd-851499a3b1d9/binary
08:52:38.572873 trace git-lfs: HTTP: 400
08:52:38.572873 trace git-lfs: HTTP: {"error":{"code":"InvalidArgumentError","title":"range first index must not be greater than last index","href":"https://api.media.atlassian.com#InvalidArgumentError"}}
```

Here's the interesting bits, the total size of the file is:

    "size": 4707098120

But then the code tries to resume from:

    Attempting to resume download of "..." from byte 4707098120

Note, that's the same number! So looks like the temp was was already complete, but the code doesn't detect that and instead tries to resume.

[The code that sets the header](https://github.com/git-lfs/git-lfs/blob/master/tq/basic_download.go#L108) then does this:

    req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", fromByte, t.Size-1))

Which results in `Range: bytes=4707098120-4707098119`, which causes "range first index must not be greater than last index".

(I can try to reproduce how we got into that state, but I think this might already be enough to solve the problem. This was with version 2.4.2 - I've looked through the release notes since then and it doesn't look like this code has changed since then.)

**Comment by ttaylorr** [comment]
My hunch is that the 'basic' adapter saw an error (perhaps even io.EOF)
when uploading/downloading the original file, and didn't bother to check
that the number of bytes transferred was its target.

In this case, I think it's probably safe to teach the adapter to log and
ignore any such errors.

> (I can try to reproduce how we got into that state, but I think this
> might already be enough to solve the problem. This was with version
> 2.4.2 - I've looked through the release notes since then and it
> doesn't look like this code has changed since then.)

I don't think that any code has changed in that tree since v2.4.2 (and
likely earlier), but it's a good thing to check before we dive in too
deeply.

Would you mind trying to reproduce this issue with the latest release
and reporting back your results here?

**Comment by robinst** [comment]
Reproduced with 2.5.0 now (and also logging of the headers):

```
10:53:57.302830 trace git-lfs: xfer: Attempting to resume download of "63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073" from byte 4707098120
10:53:57.303003 trace git-lfs: HTTP: GET https://api.media.atlassian.com/file/4d1155f8-7d31-4549-9dcd-851499a3b1d9/binary
> GET /file/4d1155f8-7d31-4549-9dcd-851499a3b1d9/binary HTTP/1.1
> Host: api.media.atlassian.com
> Authorization: Bearer ...
> Range: bytes=4707098120-4707098119
> User-Agent: git-lfs/2.5.0 (GitHub; darwin amd64; go 1.9.4)
> X-Client-Id: 6ff39d78-635f-4bd8-a06c-89dc7129dec2
```

**Comment by robinst** [comment]
> Wait, this looks like a new bug? The range is backwards in the HTTP trace output that you posted.

No, that's the bug I was describing in the issue description already:

> Which results in Range: bytes=4707098120-4707098119, which causes "range first index must not be greater than last index".

I didn't have enough scrollback before, but I've run it again and there's indeed an earlier error:

```
Expected OID 63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073, got 988666ebdcff9af9e6e7e4e8293d230d4c5ec9f9e5b1760f03581d9a308b1e31 after 3633366105 bytes written
```

Looking at the incomplete file:

```
$ ls -l .git/lfs/incomplete
-rw-r--r--  1 rstocker  staff  4707098120  9 Aug 10:14 63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073.tmp
$ gsha256sum .git/lfs/incomplete/63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073.tmp
988666ebdcff9af9e6e7e4e8293d230d4c5ec9f9e5b1760f03581d9a308b1e31
```

So the hash of the file does not match the hash in the pointer. I downloaded the file manually (without using range headers) and I get the above (incorrect) hash too instead of the expected 63a877ced56cba36aa0a83aa8824862141208e108452313384cbf591ad4b3073.

So it looks like something went wrong with uploading there. I'll try to get the customer to upload it again.

So the root cause is not a bug in git lfs :)! But what it could do is detect that case and print a better message instead of trying to make an invalid range request.

**Comment by ttaylorr** [comment]
> So the root cause is not a bug in git lfs :)!

That's good :-). I agree with what you wrote in an earlier comment that
it sounds like what we're experiencing with this behavior is a
downstream effect of a failed upload.

> But what it could do is detect that case and print a better message
> instead of trying to make an invalid range request.

I'm not sure exactly what you're looking for, besides detecting a case
like: "We were about to make an invalid Range request, and can't".

I don't think that this error makes much sense to a user without other
details (like, this file didn't match the expected OID even after we
downloaded the whole thing).

Maybe that's the error that you're looking for? Something that happens
in the case of: "We tried to download the whole thing, we did, but the
content OID didn't match. So, instead of retrying from the end (which is
nonsensical), here's a friendlier error."

**Comment by robinst** [comment]
> "We tried to download the whole thing, we did, but the content OID didn't match. So, instead of retrying from the end (which is nonsensical), here's a friendlier error."

Pretty much, yeah. That would be a good message!
